File: examples/go/clean-api-go/domain/entities/sale.go
package entities

import (
	"time"

	"github.com/google/uuid"
)

// Sale is an append-only record stored in MongoDB, one per completed sale.
// UnitPrice is a snapshot taken at sale time — it must never be recomputed
// from the current Product.Price, since that would silently rewrite history
// whenever the catalog price changes later.
type Sale struct {
	ID          string    `bson:"_id" json:"id"`
	ProductID   string    `bson:"product_id" json:"product_id"`
	Quantity    int       `bson:"quantity" json:"quantity"`
	UnitPrice   float64   `bson:"unit_price" json:"unit_price"`
	TotalAmount float64   `bson:"total_amount" json:"total_amount"`
	CreatedAt   time.Time `bson:"created_at" json:"created_at"`
}

func NewSale(productID string, quantity int, unitPrice float64) *Sale {
	return &Sale{
		ID:          uuid.New().String(),
		ProductID:   productID,
		Quantity:    quantity,
		UnitPrice:   unitPrice,
		TotalAmount: unitPrice * float64(quantity),
		CreatedAt:   time.Now().UTC(),
	}
}

File: examples/python/clean_api_python/domain/entities/sale.py
from dataclasses import dataclass, field
from datetime import datetime, timezone
from uuid import uuid4


@dataclass
class Sale:
    """Append-only record stored in MongoDB, one per completed sale.
    unit_price is a snapshot taken at sale time — never recomputed from the
    current Product.price, since that would silently rewrite history
    whenever the catalog price changes later."""

    product_id: str
    quantity: int
    unit_price: float
    id: str = field(default_factory=lambda: uuid4().hex)
    created_at: datetime = field(default_factory=lambda: datetime.now(timezone.utc))

    @property
    def total_amount(self) -> float:
        return self.unit_price * self.quantity

    def to_doc(self) -> dict:
        return {
            "_id": self.id,
            "product_id": self.product_id,
            "quantity": self.quantity,
            "unit_price": self.unit_price,
            "total_amount": self.total_amount,
            "created_at": self.created_at,
        }

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "product_id": self.product_id,
            "quantity": self.quantity,
            "unit_price": self.unit_price,
            "total_amount": self.total_amount,
            "created_at": self.created_at.isoformat(),
        }
